godot-core/src/builtin/plane.rs

```diff
@@ -0,0 +1,176 @@
+
+use super::{is_equal_approx, real, Vector3};
+
+/// 3D plane in [Hessian normal form](https://mathworld.wolfram.com/HessianNormalForm.html).
+///
+/// The Hessian form defines all points `point` which satisfy the equation
+/// `dot(normal, point) + d == 0`, where `normal` is the normal vector and `d`
+/// the distance from the origin.
+///
+/// Currently most methods are only available through [`InnerPlane`](super::inner::InnerPlane).
+///
+/// Note: almost all methods on `Plane` require that the `normal` vector have
+/// unit length and will panic if this invariant is violated. This is not separately
+/// annotated for each method.
+#[derive(Copy, Clone, PartialEq, Debug)]
+#[repr(C)]
+pub struct Plane {
+    pub normal: Vector3,
+    pub d: real,
+}
+
+impl Plane {
+    /// Creates a new `Plane` from the `normal` and the distance from the origin `d`.
+    ///
+    /// # Panics
+    /// In contrast to construction via `Plane { normal, d }`, this verifies that `normal` has unit length, and will
+    /// panic if this is not the case.
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal, float d)`_
+    #[inline]
+    pub fn new(unit_normal: Vector3, d: real) -> Self {
+        let plane = Self {
+            normal: unit_normal,
+            d,
+        };
+        plane.assert_normalized();
+        plane
+    }
+
+    /// Create a new `Plane` through the origin from a normal.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal)`_
+    #[inline]
+    pub fn from_normal_at_origin(normal: Vector3) -> Self {
+        Self::new(normal, 0.0)
+    }
+
+    /// Create a new `Plane` from a normal and a point in the plane.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(Vector3 normal, Vector3 point)`_
+    #[inline]
+    pub fn from_point_normal(point: Vector3, normal: Vector3) -> Self {
+        Self::new(normal, normal.dot(point))
+    }
+
+    /// Creates a new `Plane` from normal and origin distance.
+    ///
+    /// `nx`, `ny`, `nz` are used for the `normal` vector.
+    /// `d` is the distance from the origin.
+    ///
+    /// # Panics
+    /// See [`Self::new()`].
+    ///
+    /// _Godot equivalent: `Plane(float a, float b, float c, float d)`_
+    #[inline]
+    pub fn from_components(nx: real, ny: real, nz: real, d: real) -> Self {
+        Self::new(Vector3::new(nx, ny, nz), d)
+    }
+
+    /// Creates a new `Plane` from three points, given in clockwise order.
+    ///
+    /// # Panics
+    /// Will panic if all three points are colinear.
+    ///
+    /// _Godot equivalent: `Plane(Vector3 point1, Vector3 point2, Vector3 point3)`_
+    #[inline]
+    pub fn from_points(a: Vector3, b: Vector3, c: Vector3) -> Self {
+        let normal = (a - c).cross(a - b);
+        assert_ne!(
+            normal,
+            Vector3::ZERO,
+            "The points {a}, {b}, {c} are all colinear"
```

Comment:
Small nitpick regarding panic messages: they should not start capitalized (and not end with period), and feel free to omit any "boilerplate" that doesn't carry information. That's e.g. what [serde recommends](https://docs.rs/serde/latest/serde/de/trait.Error.html#tymethod.custom).

> "points {a}, {b}, {c} are all colinear"

Also in a few other places.